scripts/languageserver/parse.jl

```diff
@@ -0,0 +1,243 @@
+type VarInfo
+    t
+    doc::String
+end
+
+type Block
+    uptodate::Bool
+    ex::Any
+    range::Range
+    name::String
+    var::VarInfo
+    localvar::Dict{String,VarInfo}
+    diags::Vector{Diagnostic}
+end
+
+function Block(utd, ex, r::Range)
+    t, name, doc, lvars = classify_expr(ex)
+    ctx = LintContext()
+    ctx.lineabs = r.start.line+1
+    dl = r.end.line-r.start.line-ctx.line
+    Lint.lintexpr(ex, ctx)
+    diags = map(ctx.messages) do l
+        return Diagnostic(Range(Position(r.start.line+l.line+dl-1, 0), Position(r.start.line+l.line+dl-1, 100)),
+                        LintSeverity[string(l.code)[1]],
+                        string(l.code),
+                        "Lint.jl",
+                        l.message) 
+    end
+    v = VarInfo(t, doc)
+
+    return Block(utd, ex, r, name,v, lvars, diags)
+end
+
+function Base.parse(uri::String, server::LanguageServer, updateall=false)
+    doc = String(server.documents[uri].data)
+    linebreaks = get_linebreaks(doc) 
+    n = length(doc.data)
+    if doc==""
+        server.documents[uri].blocks = []
+        return
+    end
+    i = findfirst(b->!b.uptodate, server.documents[uri].blocks)
+
+    if isempty(server.documents[uri].blocks) || updateall || i==0
+        i0 = i1 = 1
+        p0 = p1 = Position(0, 0)
+        out = Block[]
+        i4 = 0
+    else
+        i4 = findnext(b->b.uptodate, server.documents[uri].blocks,i)
+        p0 = p1 = server.documents[uri].blocks[i].range.start
+        i0 = i1 = linebreaks[p0.line+1]+p0.character+1
+        out = server.documents[uri].blocks[1:i-1]
+    end
```

Comment:
Maybe add some comments here? Really difficult to understand what exactly is going on... Would probably be enough to have one comment at the top of the block that gives a hint what this is about, and then maybe comments for the various vars that describe what they mean.